### PR TITLE
fix(aio): support IE via conditionally loaded polyfills

### LINKED ISSUE #16519: aio: IE no bueno. Need polyfill strategy
Angular apps - including this one - won't run in IE without polyfill. But a polyfill for everyone will slow loading. What do we do? We don't want to punish everyone to make IE happy. @IgorMinar - a solution please?

Btw, the fact that our aio code uses features not native to IE is a drop in the bucket.

### PR DESCRIPTION
**Pull request by petebacondarwin**
It was not possible to use a request directly to a service like `polyfill.io` since it pulled down additional polyfills that broke other aspects of Angular/Zones. Instead we are generating our own IE polyfill file from the `core-js` and other libraries directly using webpack.

It was not possible to use HTML conditional comments to include the script tag for loading the IE polyfill, so we are using a small script block that checks for `window.document.documentMode` and then adds the script tag that loads the polyfill to the end of the `<head>`. Although, in the developer console it appears that the main Angular source code is being downloaded first, they are not run until this polyfill has run.

If we are not running inside IE then this file is never downloaded, and there does not appear to be a difference in the speed at which the application downloads and boots.

To test this out on IE, don't forget to run `yarn setup` (or just `yarn build-ie-polyfills` to generate the IE polyfill file, before starting the web server. I found that disabling the live-reload made things easier; and setting the host to 0.0.0.0 made it easier to work with running Windows in a VM.

```
yarn start -- --host 0.0.0.0 --live-reload false
```

Closes #16519

### CODE REVIEW

File: aio/src/index.html [hunk 1]
```diff
@@ -23,7 +23,7 @@
   <!-- NOTE: These need to be kept in sync with `ngsw-manifest.json`. -->
   <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
   <link href="https://fonts.googleapis.com/css?family=Droid+Sans+Mono" rel="stylesheet">
-  <link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css"rel="stylesheet">
+  <link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" rel="stylesheet">
```
Comment (gkalpak) [inline aio/src/index.html]: Yeah, IE complains about stuff like this 😛

File: aio/src/ie-polyfills.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import 'core-js/es6/array';
+import 'core-js/es6/regexp';
+import 'core-js/es6/map';
+import 'core-js/es6/set';
+
+/** IE10 and IE11 requires the following for NgClass support on SVG elements */
+import 'classlist.js';
+
+/** IE10 and IE11 requires the following to support `@angular/animation`. */
+import 'web-animations-js';
```
Comment (gkalpak) [inline aio/src/ie-polyfills.ts]: Why are there both `.ts` and `.js` versions of the `ie-polyfills` file?

File: aio/package.json [hunk 3]
```diff
@@ -33,7 +33,8 @@
     "generate-plunkers": "node ./tools/plunker-builder/generatePlunkers",
     "generate-zips": "node ./tools/example-zipper/generateZips",
     "sw-manifest": "ngu-sw-manifest --dist dist --in ngsw-manifest.json --out dist/ngsw-manifest.json",
-    "sw-copy": "cp node_modules/@angular/service-worker/bundles/worker-basic.min.js dist/"
+    "sw-copy": "cp node_modules/@angular/service-worker/bundles/worker-basic.min.js dist/",
+    "build-ie-polyfills": "webpack -p src/ie-polyfills.js src/ie-polyfills.min.js"
```
Comment (gkalpak) [inline aio/package.json]: Having generated content in `src/` makes me 😢 

Why not move this step after `ng build` and put it into `dist/` (similar to `sw-manifest`/`sw-copy`)?
There could be a `postbuild` task that does all three.

File: aio/README.md [hunk 4]
```diff
@@ -30,6 +30,7 @@ Here are the most important tasks you might need to use:
 * `yarn generate-plunkers` - generate the plunker files that are used by the `live-example` tags in the docs.
 * `yarn generate-zips` - generate the zip files from the examples. Zip available via the `live-example` tags in the docs.
 
+* `yarn build-ie-polyfills` - generates a js file of polyfills that can be loaded in Internet Explorer
```
Comment (gkalpak) [inline aio/README.md]: Missing trailing `.` (for consistency) :grin:

**Review by gkalpak — COMMENTED** [review]
Some pages seem to trigger full page reloads (e.g. [this one][1]) and the links in the footer are not clickable.
Other than that (and some styles being off - which is unavoidable I think), it is surprisingly functional 😄 

[1]: https://pr16805-23553e3e9920301a089cbec72a0cda57a38b909b.ngbuilds.io

File: aio/src/ie-polyfills.ts (hunk 2 above)
Comment (petebacondarwin) [inline aio/src/ie-polyfills.ts]: I messed up a commit amend - there should only be a .js version

File: aio/package.json (hunk 3 above)
Comment (petebacondarwin) [inline aio/package.json]: I don't think that would work with the webpack development server :-(

File: aio/README.md (hunk 4 above)
Comment (petebacondarwin) [inline aio/README.md]: +1

File: aio/package.json (hunk 3 above)
Comment (gkalpak) [inline aio/package.json]: It won't 😞 
<sub>(which I can live with, but don't feel too strongly about)/sub>

Comment (petebacondarwin) [inline aio/package.json]: How about we move the file into the `generated` folder? At least then it is not going to be confused with "real" source code?

Comment (gkalpak) [inline aio/package.json]: 💯

**Comment by petebacondarwin** [comment]
Generated code moved to .... `generated`. PTAL @gkalpak

File: aio/src/NOTES.md [hunk 5]
```diff
@@ -1,159 +0,0 @@
...
-
-
-```ts
-class NavigationService {
-  navigationViews: Observable<NavigationViews>;
-  currentNode: Observable<NavigationNode>;
-  activeNodes: Observable<NavigationNode[]>;
-}
-```
-
```
Comment (petebacondarwin) [inline aio/src/NOTES.md]: This whole file is completely out of date and so is actually acting as a red herring for anyone who wants to understand the code.

**Comment by petebacondarwin** [comment]
In keeping with the yarn dependencies theme, I expect that @IgorMinar will want to land his big PR before this one.

**Comment by gkalpak** [comment]
The changes look good. It depends on what level of functionality we are after (see https://github.com/angular/angular/pull/16805#pullrequestreview-38323684).

**Comment by petebacondarwin** [comment]
I was aiming for "basic" functionality :-) as a starter.

**Review by gkalpak — APPROVED** [review]
LGTM

There are still few issues, but we can work on them later.

**Pull request merged**